### PR TITLE
Lstebner/achievement i am rich

### PR DESCRIPTION
**Pull request by lstebner**
This PR implements 3 new achievements in a series referred to as "I am Rich". this closes issue #73 , and can be tested here https://farmhand-git-lstebner-achievement-i-am-rich-jeremyckahn.vercel.app/

this PR also adds a few new utils which were used to implement this feature

- `percentageString` util to convert a floating percent to a string value
- `getSalePriceMultiplier` looks at achievements and returns a multiplier value

completed achievements:

Screen Shot 2021-05-29 at 2 37 17 PM

incomplete achievements:

Screen Shot 2021-05-30 at 9 35 01 AM

increased price in sell view:

Screen Shot 2021-05-29 at 2 34 28 PM

unit tests:

Screen Shot 2021-05-29 at 1 32 34 PM

Screen Shot 2021-05-29 at 1 33 03 PM

### CODE REVIEW

File: src/reducers.js [hunk 1]
```diff
@@ -1129,6 +1129,17 @@ export const sellItem = (state, { id }, howMany = 1) => {
...
+
+  if (state.completedAchievements['i-am-rich-3']) {
+    salePriceMultiplier = 1.25
+  } else if (state.completedAchievements['i-am-rich-2']) {
+    salePriceMultiplier = 1.1
+  } else if (state.completedAchievements['i-am-rich-3']) {
+    salePriceMultiplier = 1.05
+  }
+
+  console.log('price multiplier value', salePriceMultiplier)
```
Comment (jeremyckahn) [inline src/reducers.js:1142]: I know you're aware of this, I just want to add this comment so we don't forget to remove it before merging! 🙂

File: src/reducers.js [hunk 2]
```diff
@@ -1129,6 +1129,17 @@ export const sellItem = (state, { id }, howMany = 1) => {
...
   let saleValue = 0
+  let salePriceMultiplier = 1
+
+  if (state.completedAchievements['i-am-rich-3']) {
+    salePriceMultiplier = 1.25
+  } else if (state.completedAchievements['i-am-rich-2']) {
+    salePriceMultiplier = 1.1
+  } else if (state.completedAchievements['i-am-rich-3']) {
+    salePriceMultiplier = 1.05
+  }
```
Comment (jeremyckahn) [inline src/reducers.js:1140]: This comment is aimed at your question in the `getItemCurrentValue` util, but makes more sense to answer here.

> should valueAdjustments have a "global" adjustment or something for any permanent global increases?

We should avoid using `valueAdjustments` to calculate sale bonuses, as that object is intended to only define market fluctuations. With that said, my intuition is to move this `salePriceMultiplier` calculation to a helper function in `src/utils.js` and use that helper both here and in `getItemCurrentValue`. What do you think of that approach?

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
This is off to a great start! I added comments throughout, but let me know if anything is unclear and we can discuss further.

File: src/data/achievements.js [hunk 3]
```diff
@@ -241,6 +241,33 @@ const achievements = [
       inventoryLimit: state.inventoryLimit + reward,
     }),
   }))(),
+
+  ((goal = 500000) => ({
+    id: 'i-am-rich-1',
+    name: 'I am Rich!',
+    description: `Earn ${dollarString(goal)}`,
+    rewardDescription: 'All sales get a 5% bonus',
+    condition: state => state.revenue >= goal,
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/data/achievements.js:250]: i scoped out the 'stats' window to see what that was using for a hint

File: src/data/achievements.js [hunk 4]
```diff
@@ -241,6 +241,33 @@ const achievements = [
       inventoryLimit: state.inventoryLimit + reward,
     }),
   }))(),
+
+  ((goal = 500000) => ({
+    id: 'i-am-rich-1',
+    name: 'I am Rich!',
+    description: `Earn ${dollarString(goal)}`,
+    rewardDescription: 'All sales get a 5% bonus',
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/data/achievements.js]: makes sense! i made a hash of `I_AM_RICH_BONUSES`, the keys line up to the # used in the 'id', which I thought was simpler to understand than just an array, but am open to whatever you prefer here.

File: src/data/achievements.test.js [hunk 5]
```diff
@@ -43,3 +43,42 @@ describe('harvest-crop', () => {
...
+      id: 'i-am-rich-2',
+      goal: { revenue: 1000000 },
+    },
+    {
+      id: 'i-am-rich-3',
+      goal: { revenue: 1000000000 },
+    },
+  ]
+
+  for (let variant of variants) {
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/data/achievements.test.js:63]: today i learned about `describe.each`. looks perfect

[the remark this answers is not in the document]
Comment (lstebner) [reply src/data/achievements.test.js]: updated those

File: src/reducers.js [hunk 6]
```diff
@@ -1129,6 +1129,17 @@ export const sellItem = (state, { id }, howMany = 1) => {
     : getAdjustedItemValue(valueAdjustments, id)
   const saleIsGarnished = isItemAFarmProduct(item)
   let saleValue = 0
+  let salePriceMultiplier = 1
+
+  if (state.completedAchievements['i-am-rich-3']) {
+    salePriceMultiplier = 1.25
+  } else if (state.completedAchievements['i-am-rich-2']) {
+    salePriceMultiplier = 1.1
+  } else if (state.completedAchievements['i-am-rich-3']) {
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/reducers.js]: good catch!

File: src/reducers.js (hunk 1 above)
Comment (lstebner) [reply src/reducers.js:1143]: removing this now

File: src/reducers.js (hunk 2 above)
Comment (lstebner) [reply src/reducers.js:1141]: i had a feeling `valueAdjustments` was something market specific. can we update the variable name? (is that easy? to like, `marketValueAdjustments`?). or add some documentation around it. i didn't see any explanation for the variable anywhere. whatever you prefer @jeremyckahn . i can also do the work if you let me know which you prefer.

onto the util...that makes sense to me! that will scale nicely later on.

Comment (lstebner) [reply src/reducers.js]: lots of unit tests, and a util or two later i think i got there!

File: src/data/achievements.test.js [hunk 7]
```diff
@@ -43,3 +43,47 @@ describe('harvest-crop', () => {
...
 })
+
+const iAmRichVariants = [
+  ['i-am-rich-1', 500000, 'Earn $500,000', 'All sales receive a 5% bonus.'],
+  ['i-am-rich-2', 1000000, 'Earn $1,000,000', 'All sales receive a 10% bonus.'],
+  [
+    'i-am-rich-3',
+    1000000000,
+    'Earn $1,000,000,000',
+    'All sales receive a 25% bonus.',
```
Comment (lstebner) [inline src/data/achievements.test.js:54]: not sure how you feel about these kinda copy tests, they could also be snapshots @jeremyckahn . i kept manually testing that the copy was correct though so it felt right to add an automated test for it

File: src/components/Item/Item.js [hunk 8]
```diff
@@ -145,6 +146,8 @@ export const Item = ({
 
   const avatar = <img {...{ src: items[id] }} alt={name} />
 
+  adjustedValue *= getSalePriceMultiplier(completedAchievements)
```
Comment (jeremyckahn) [inline src/components/Item/Item.js:149]: Rather than mutating `adjustedValue` here, I think we're going to need to add an additional number to represent `adjustedValue * getSalePriceMultiplier(completedAchievements)`. The reason for this is that `adjustedValue`, at least for non-Shop items, specifically defines the value of an item adjusted for current market values. Modifying it here will break the calculations at these points deeper in the component:

https://github.com/jeremyckahn/farmhand/blob/26d249bac2678abda5e3417d951ff69224632d92/src/components/Item/Item.js#L204-L206

https://github.com/jeremyckahn/farmhand/blob/26d249bac2678abda5e3417d951ff69224632d92/src/components/Item/Item.js#L233-L235

This is a bit confusing because #73 makes the semantics of this variable a bit unclear. It may make sense to rename `adjustedValue` to something like `marketAdjustedValue`, since the current name is too vague now. I think "adjustedValue" is good for representing "the final number that is presented to the user," in case you need some help with the variable naming here.

File: src/utils.js [hunk 9]
```diff
@@ -192,6 +193,12 @@ export const dollarString = number => formatNumber(number, '$0,0')
  */
 export const integerString = number => formatNumber(number, '0,0')
 
+/**
+ * @param {number} a float
```
Comment (jeremyckahn) [inline src/utils.js:197]: Per JSDoc syntax, this needs to reference the name of the variable. So, this will need to be something like:

```
* @param {number} number A float
```

File: src/utils.js [hunk 10]
```diff
@@ -989,3 +997,21 @@ export const getCostOfNextStorageExpansion = currentInventoryLimit => {
...
+  if (completedAchievements['i-am-rich-3']) {
+    salePriceMultiplier += I_AM_RICH_BONUSES[3]
+  } else if (completedAchievements['i-am-rich-2']) {
+    salePriceMultiplier += I_AM_RICH_BONUSES[2]
+  } else if (completedAchievements['i-am-rich-1']) {
+    salePriceMultiplier += I_AM_RICH_BONUSES[1]
+  }
+
+  return salePriceMultiplier
+}
```
Comment (jeremyckahn) [inline src/utils.js:1017]: Love it!

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Awesome progress so far! There's a few issues that need to be addressed, but I'm really excited to see how well this is coming along!

File: src/reducers.js (hunk 2 above)
Comment (jeremyckahn) [reply src/reducers.js]: > i had a feeling `valueAdjustments` was something market specific. can we update the variable name? (is that easy? to like, `marketValueAdjustments`?). or add some documentation around it. i didn't see any explanation for the variable anywhere. whatever you prefer @jeremyckahn . i can also do the work if you let me know which you prefer.

Whoops, I saw this after the fact! Sorry about that. I've effectively addressed this here: https://github.com/jeremyckahn/farmhand/pull/77/files#r641988878

tl;dr: Yes, we should definitely rename `valueAdjustments` to something like `marketValueAdjustments` for the relative value calculations, but we'll also need to having some sort of "grand total" value that is presented to the user as well.

File: src/data/achievements.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/data/achievements.js]: I noticed this reply late, but I addressed it in https://github.com/jeremyckahn/farmhand/pull/77/files#r641982208, but yes I (personally) think it would a little more intuitive for this to be an `Array`. 🙂

File: src/components/Item/Item.js (hunk 8 above)
Comment (lstebner) [reply src/components/Item/Item.js:149]: i see now, thanks for pointing that out. i think it actually makes the most sense to add a `sellPrice` variable and only adjust the "Sell price" field. i would like to see `adjustedValue` updated as you said, but there are many other similar variables that i would also want to update like `previousDatAdjustedValue` and `historicalValueAdjustments` so I think it's best to leave those all for now. how does that sound?

File: src/utils.js (hunk 9 above)
Comment (lstebner) [reply src/utils.js:197]: oops, number/number! good catch

File: src/data/achievements.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (lstebner) [reply src/data/achievements.js]: sounds good! going to resolve this conversation since that one exists now

File: src/data/achievements.test.js (hunk 7 above)
Comment (lstebner) [reply src/data/achievements.test.js:54]: sounds good to me!

File: src/data/achievements.js [hunk 11]
```diff
@@ -241,6 +242,40 @@ const achievements = [
...
     }),
   }))(),
+
+  ((goal = 500000) => ({
+    id: 'i-am-rich-1',
+    name: 'I am Rich!',
+    description: `Earn ${dollarString(goal)}`,
+    rewardDescription: `All sales receive a ${percentageString(
+      I_AM_RICH_BONUSES[1],
+      '%'
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/data/achievements.js:252]: yup! a formatNumber attempt hah. thanks for noticing that

**Comment by lstebner** [comment]
@jeremyckahn pushed another round of updates for all your latest comments

File: src/components/Item/Item.js [hunk 12]
```diff
@@ -145,6 +146,9 @@ export const Item = ({
 
   const avatar = <img {...{ src: items[id] }} alt={name} />
 
+  const sellPrice =
```
Comment (jeremyckahn) [inline src/components/Item/Item.js:149]: This is exactly what I was thinking! I think we're going to need `sellPrice` instead of `adjustedValue` here on like 294 as well: https://github.com/jeremyckahn/farmhand/blob/cea1cb22029552865edf3a3f5c45556e57dd9f20/src/components/Item/Item.js#L288-L300

Once that's fixed, I think this should be all good! 🙂

File: src/constants.js [hunk 13]
```diff
@@ -110,6 +110,8 @@ export const DAILY_FINANCIAL_HISTORY_RECORD_LENGTH = 7
 
 export const RECIPE_INGREDIENT_VALUE_MULTIPLIER = 1.25
 
+export const I_AM_RICH_BONUSES = [0.05, 0.1, 0.25]
```
Comment (jeremyckahn) [inline src/constants.js:113]: I love this! Thank for for making this change! 🙏

File: src/data/achievements.js [hunk 14]
```diff
@@ -241,6 +242,39 @@ const achievements = [
       inventoryLimit: state.inventoryLimit + reward,
     }),
   }))(),
+
+  ((goal = 500000) => ({
+    id: 'i-am-rich-1',
+    name: 'I am Rich!',
+    description: `Earn ${dollarString(goal)}`,
```
Comment (jeremyckahn) [inline src/data/achievements.js:249]: Sorry I missed this previously, but these `description` strings will also need a period at the end to be consistent with the other achievements.

File: src/utils.js [hunk 15]
```diff
@@ -204,6 +211,7 @@ const getItemBaseValue = itemId => itemsMap[itemId].value
  * @returns {number}
  */
 export const getItemCurrentValue = ({ id }, valueAdjustments) =>
+  // should valueAdjustments have a "global" adjustment or something for any permanent global increases?
```
Comment (jeremyckahn) [inline src/utils.js:214]: Since it seems like we're getting close to the end with this PR, please be sure to remove this comment (unless it's something you feel is unresolved). 🙂

File: src/utils.js [hunk 16]
```diff
@@ -232,7 +240,8 @@ export const isItemSoldInShop = ({ id }) => Boolean(shopInventoryMap[id])
  * @param {farmhand.item} item
  * @returns {number}
  */
-export const getResaleValue = ({ id }) => itemsMap[id].value / 2
+export const getResaleValue = ({ id }, completedAchievements) =>
```
Comment (jeremyckahn) [inline src/utils.js:243]: It looks like we can drop the `completedAchievements` parameter here.

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Thanks for the latest iteration! This is looking great and I think we're almost there. I saw a few mostly minor regression issues to address, but I think the hard problems are solved at this point!

File: src/components/Item/Item.js (hunk 8 above)
Comment (jeremyckahn) [reply src/components/Item/Item.js]: Oh yeah, those variables are functionally affected by this feature at all, and changing their name is actually quite a bit more involved because they are part of the persisted game data. They're named slightly more vaguely than I'd like right now, but it's not worth changing them in this PR.

File: src/components/Item/Item.js (hunk 12 above)
Comment (lstebner) [reply src/components/Item/Item.js:149]: oh wow i did not see that one at all!

File: src/utils.js (hunk 15 above)
Comment (lstebner) [reply src/utils.js:214]: another good catch! we resolved this with the discussion about these being only for market adjustments 👍

File: src/utils.js (hunk 16 above)
Comment (lstebner) [reply src/utils.js:243]: oops! yep, didn't end up going that route. removing

File: src/components/Item/Item.js (hunk 12 above)
Comment (lstebner) [reply src/components/Item/Item.js:149]: actually, i'm not sure this one was supposed to change 🤔 . this is for purchasing, right? we only want to affect sell price I think

Comment (jeremyckahn) [reply src/components/Item/Item.js:149]: Ah! Great catch, you're right. Thanks for scrutinizing the code extra closely here! 😅

**Comment by jeremyckahn** [comment]
The code looks good to me, but I just tried to test out the new achievement rewards and they don't seem to factor in the sales bonus. 😕

Here's where I tested out the new feature: https://farmhand-git-lstebner-achievement-i-am-rich-jeremyckahn.vercel.app/

Here's the save file I used to test this out: https://gist.github.com/jeremyckahn/4dc3bd35f6e2db3c47b24dcc62a5db27

(You can save that JSON data to a local file and then import it into the game for testing.)

You'll see there that Carrots have a boosted value of $26.25, but once sold, the player only gets the non-boosted value of $25. I'll take a look to see what part of the code might need to be revisited to fix this too.

**Comment by jeremyckahn** [comment]
> The code looks good to me, but I just tried to test out the new achievement rewards and they don't seem to factor in the sales bonus. 😕
> 
> Here's where I tested out the new feature: https://farmhand-git-lstebner-achievement-i-am-rich-jeremyckahn.vercel.app/
> 
> Here's the save file I used to test this out: https://gist.github.com/jeremyckahn/4dc3bd35f6e2db3c47b24dcc62a5db27
> 
> (You can save that JSON data to a local file and then import it into the game for testing.)
> 
> You'll see there that Carrots have a boosted value of $26.25, but once sold, the player only gets the non-boosted value of $25. I'll take a look to see what part of the code might need to be revisited to fix this too.

Based on some quick experimentation, it looks like this change might be all that's needed here: 

```diff
diff --git a/src/reducers.js b/src/reducers.js
index a10d58a..617bc03 100644
--- a/src/reducers.js
+++ b/src/reducers.js
@@ -1142,12 +1142,11 @@ export const sellItem = (state, { id }, howMany = 1) => {
           castToMoney(adjustedItemValue * LOAN_GARNISHMENT_RATE)
         )
       : 0
-    const garnishedProfit = adjustedItemValue - loanGarnishment
+    const garnishedProfit =
+      adjustedItemValue * getSalePriceMultiplier(completedAchievements) -
+      loanGarnishment
     loanBalance = moneyTotal(loanBalance, -loanGarnishment)
-    saleValue = moneyTotal(
-      saleValue * getSalePriceMultiplier(completedAchievements),
-      garnishedProfit
-    )
+    saleValue = moneyTotal(saleValue, garnishedProfit)
   }
 
   if (saleIsGarnished) {
```

The tests pass with this, and I'm seeing the correct behavior in-game with it!

**Review by jeremyckahn — APPROVED** [review]
I've QA'ed and verified this on https://farmhand-git-lstebner-achievement-i-am-rich-jeremyckahn.vercel.app/, and everything seems to be working perfectly! Thank you @lstebner for sticking with this and bringing it through to completion. This feature adds some much-needed late-game content and lays the groundwork for more interesting mechanics down the line. Awesome work! LGTM! 🚀
